In [32]:
from datetime import date
import pandas as pd

pd.set_option('display.max_columns', 350)

In [33]:
import os
def get_project_root_path():
    project_directory = "fifa22-players-analysis"
    separated_path = os.getcwd().split(os.sep)
    project_directory_index = separated_path.index(project_directory) + 1
    return os.sep.join(separated_path[0:project_directory_index])


import joblib
def save_model(model, model_name):
    filename = f"model_name_{model_name}_{str(date.today())}.pkl"
    print(filename)
    model_path = os.sep.join([get_project_root_path(), "saved_models", filename])
    joblib.dump(model, model_path)

In [34]:
data = pd.read_csv(f"data/processed_data_{date.today()}.tsv", sep="\t")
data.head()

/tmp/ipykernel_108936/2624993333.py:1: DtypeWarning: Columns (22) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv(f"data/processed_data_{date.today()}.tsv", sep="\t")


sofifa_id                                         player_url  short_name  \
0     167575  https://sofifa.com/player/167575/anton-shunin/...   A. Shunin   
1     205211  https://sofifa.com/player/205211/milan-borjan/...   M. Borjan   
2     244700  https://sofifa.com/player/244700/yuriy-dyupin/...   Y. Dyupin   
3     199904  https://sofifa.com/player/199904/jesse-joronen...  J. Joronen   
4     240225  https://sofifa.com/player/240225/matvey-safono...  M. Safonov   

             long_name player_positions  overall  potential  value_eur  \
0         Anton Shunin               GK       76         76        NaN   
1         Milan Borjan               GK       76         76        NaN   
2         Yuriy Dyupin               GK       74         74        NaN   
3  Jesse Pekka Joronen               GK       72         72        NaN   
4       Matvey Safonov               GK       72         82        NaN   

   wage_eur         dob  weight_kg  height_cm  club_team_id club_position  \
0       NaN  1987-01-27         88        191           NaN           NaN   
1       NaN  1987-10-23         84        196           NaN           NaN   
2       NaN  1988-03-17         84        187           NaN           NaN   
3       NaN  1993-03-21         89        197           NaN           NaN   
4       NaN  1999-02-25         82        192           NaN           NaN   

   club_jersey_number preferred_foot  weak_foot  skill_moves  \
0                 NaN           Left          3            1   
1                 NaN          Right          2            1   
2                 NaN          Right          3            1   
3                 NaN          Right          3            1   
4                 NaN          Right          2            1   

   international_reputation      work_rate      body_type  release_clause_eur  \
0                         1  Medium/Medium    Lean (185+)                 NaN   
1                         1  Medium/Medium  Normal (185+)                 NaN   
2                         1  Medium/Medium  Normal (185+)                 NaN   
3                         1  Medium/Medium  Normal (185+)                 NaN   
4                         1  Medium/Medium  Normal (185+)                 NaN   

  player_tags                         player_traits  pace  shooting  passing  \
0         NaN                           Team Player   NaN       NaN      NaN   
1         NaN         Leadership, Comes For Crosses   NaN       NaN      NaN   
2         NaN                                   NaN   NaN       NaN      NaN   
3         NaN  GK Long Throw, Cautious With Crosses   NaN       NaN      NaN   
4         NaN                                   NaN   NaN       NaN      NaN   

   dribbling  defending  physic  attacking_crossing  attacking_finishing  \
0        NaN        NaN     NaN                  18                   11   
1        NaN        NaN     NaN                  15                   12   
2        NaN        NaN     NaN                  12                    6   
3        NaN        NaN     NaN                  15                   13   
4        NaN        NaN     NaN                  13                    7   

   attacking_heading_accuracy  skill_dribbling  skill_curve  \
0                          12               14           15   
1                          24               28           16   
2                          10                7           14   
3                          18               16           14   
4                          14               10           11   

   skill_fk_accuracy  skill_long_passing  skill_ball_control  \
0                 13                  30                  20   
1                 16                  48                  23   
2                 11                  22                  12   
3                 12                  25                  16   
4                 12                  26                  18   

   movement_acceleration  movement_sprint_spee

In [35]:
def count_na(df: pd.DataFrame) -> None:
    nan_sum = df.isna().sum()
    display(nan_sum[nan_sum > 0])

count_na(data)

value_eur                74
wage_eur                 61
club_team_id             61
club_position            61
club_jersey_number       61
release_clause_eur     1176
player_tags           17798
player_traits          9841
pace                   2132
shooting               2132
passing                2132
dribbling              2132
defending              2132
physic                 2132
goalkeeping_speed     17107
dtype: int64

Analisando essa tabela com os valores nulos podemos perceber que alguns jogadores do dataset não possuem um time atual, isso explica
`wage_eur`, `club_team_id`, `club_position`, `club_jersey_number` terem a mesma quantidade de valores nulos (61),
porém `value_eur` deveria seguir esse mesmo padrão, assim podemos dizer que 13 (74 - 61) jogadores que estão ativos não possuem um valor,
portanto precisamos prever esses dados.

Também podemos perceber que temos 2132 jogadores não possuem `pace`, `shooting`, `passing`, `dribbling`, `defending`, `physic`. Vamos analisar esses jogadores

In [36]:
def percent_of_gk(df: pd.DataFrame) -> None:
    gk = df[df['GK'] == 1]
    display(gk.shape[0] / df.shape[0])


missing_main_attr_data = data[data.pace.isna()]

percent_of_gk(missing_main_attr_data)
missing_main_attr_data

1.0

sofifa_id                                         player_url  \
0        167575  https://sofifa.com/player/167575/anton-shunin/...   
1        205211  https://sofifa.com/player/205211/milan-borjan/...   
2        244700  https://sofifa.com/player/244700/yuriy-dyupin/...   
3        199904  https://sofifa.com/player/199904/jesse-joronen...   
4        240225  https://sofifa.com/player/240225/matvey-safono...   
...         ...                                                ...   
2127     257618  https://sofifa.com/player/257618/armin-gremsl/...   
2128     262424  https://sofifa.com/player/262424/sorin-mogosan...   
2129     262420  https://sofifa.com/player/262420/robert-popa/2...   
2130     210022  https://sofifa.com/player/210022/philip-sprint...   
2131     244417  https://sofifa.com/player/244417/julian-krahl/...   

       short_name            long_name player_positions  overall  potential  \
0       A. Shunin         Anton Shunin               GK       76         76   
1       M. Borjan         Milan Borjan               GK       76         76   
2       Y. Dyupin         Yuriy Dyupin               GK       74         74   
3      J. Joronen  Jesse Pekka Joronen               GK       72         72   
4      M. Safonov       Matvey Safonov               GK       72         82   
...           ...                  ...              ...      ...        ...   
2127    A. Gremsl         Armin Gremsl               GK       63         66   
2128  S. Mogoșanu       Sorin Mogoșanu               GK       58         60   
2129      R. Popa          Robert Popa               GK       56         75   
2130    P. Sprint        Philip Sprint               GK       62         64   
2131     J. Krahl         Julian Krahl               GK       62         76   

      value_eur  wage_eur         dob  weight_kg  height_cm  club_team_id  \
0           NaN       NaN  1987-01-27         88        191           NaN   
1           NaN       NaN  1987-10-23         84        196           NaN   
2           NaN       NaN  1988-03-17         84        187           NaN   
3           NaN       NaN  1993-03-21         89        197           NaN   
4           NaN       NaN  1999-02-25         82        192           NaN   
...         ...       ...         ...        ...        ...           ...   
2127   475000.0    2000.0  1994-08-13         85        190      115716.0   
2128   170000.0    1000.0  1992-11-22         77        180      115716.0   
2129   325000.0     500.0  2003-03-05         83        190      115716.0   
2130   350000.0     700.0  1993-06-27         98        196      115820.0   
2131   850000.0     500.0  2000-01-22         90        194      115820.0   

     club_position  club_jersey_number preferred_foot  weak_foot  skill_moves  \
0              NaN                 NaN           Left          3            1   
1              NaN                 NaN          Right          2            1   
2              NaN                 NaN          Right          3            1   
3              NaN                 NaN          Right          3            1   
4              NaN                 NaN          Right          2            1   
...            ...                 ...            ...        ...          ...   
2127           RES                94.0           Left          2            1   
2128            GK                 1.0          Right          1            1   
2129           SUB                31.0          Right          2            1   
2130            GK                 1.0          Right          2            1   
2131           SUB                13.0          Right          3            1   

      international_reputation      work_rate         body_type  \
0                            1  Medium/Medium       Lean (185+)   
1                            1  Medium/Medium     Normal (185+)   
2                            1  Medium/Medium     Normal (185+)   
3                            1  Medium/Medium     

Como vimos que os goleiros não tem os atributos listados acima, mas possuem um overall em cada posição, seria interessante prever esses atributos.

In [37]:
missing_gk_speed_data = data[data.goalkeeping_speed.isna()]
sum_of_missing_data = missing_gk_speed_data.shape[0] + missing_main_attr_data.shape[0]
print(f"Jogadores sem gk speed + jogadores sem atributos primários: {sum_of_missing_data}")
print(f"Numero de jogadores: {data.shape[0]}")

Jogadores sem gk speed + jogadores sem atributos primários: 19239
Numero de jogadores: 19239


Assim podemos ver que todos os jogadores que não são goleiros não possuem `goalkeeping_speed`

## Prevendo valor dos jogadores
Nessa etapa vamos criar um modelo para prever o valor dos 13 jogadores que não possuem valor

In [38]:
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
import numpy as np

def predict_and_populate(df_to_populate, df_to_train, model, column_to_predict, round_=None, save=False) -> pd.DataFrame:
    # Separating the target variable
    target = df_to_train[column_to_predict]
    features = df_to_train.drop(column_to_predict, axis=1)

    # Poluting the missing features
    mean_imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
    mean_imputer.fit(features)
    imputed_features = mean_imputer.transform(features)

    # Train test split
    features_train, features_test, target_train, target_test = train_test_split(imputed_features, target, test_size=0.2)

    # Training the model
    trained_model = model.fit(features_train, target_train)

    # Saving the model
    if save:
        model_name = str(model).split("(")[0]
        filename = f"{model_name}_{column_to_predict}"
        save_model(trained_model, filename)

    # Calculatin model score
    model_score = trained_model.score(features_test, target_test)
    print(f"Score do modelo: {model_score:.4f}")

    # Populating the missing values
    feature_to_populate = mean_imputer.transform(df_to_populate.drop(column_to_predict, axis=1))
    predicted_values = trained_model.predict(feature_to_populate)

    if round_ is not None:
        predicted_values = predicted_values.round(round_)

    df_to_populate[column_to_predict] = predicted_values

    return df_to_populate

In [39]:
players_with_club_and_no_value = data.query("not club_team_id.isna() and value_eur.isna()").copy()
data_without_players_with_club_and_no_value = data.drop(players_with_club_and_no_value.index)

data_to_train = data_without_players_with_club_and_no_value.query("not value_eur.isna()").copy()
columns_to_drop = [
    "sofifa_id", "player_url", "short_name", "long_name",
    "dob", "club_position","player_traits", "player_tags",
    "player_positions", "work_rate", "body_type",
    "preferred_foot", "nationality_name"
]

data_to_train = data_to_train.drop(columns=columns_to_drop)
data_to_fill = players_with_club_and_no_value.drop(columns=columns_to_drop)

In [40]:
# Normalizando os dados
# from sklearn.preprocessing import StandardScaler
#
# standerd_scaler =  StandardScaler()
# standardized_features = standerd_scaler.fit_transform(imputed_features)
# standardized_features

## Comparando modelos diferentes

In [41]:
# TODO ver tanto o MSE quanto o MAE
# TODO aplicar RFECV em cada modelo

from sklearn.dummy import DummyRegressor

dummy = DummyRegressor()
predict_and_populate(data_to_fill, data_to_train, dummy, "value_eur");

Score do modelo: -0.0003


In [42]:
from sklearn.linear_model import LinearRegression

linear_regression = LinearRegression()
predict_and_populate(data_to_fill, data_to_train, linear_regression, "value_eur");

Score do modelo: 0.9733


In [43]:
from sklearn.ensemble import RandomForestRegressor

random_forest = RandomForestRegressor(random_state=0, n_jobs=-1, oob_score=True)
players_with_club_and_no_value = predict_and_populate(
    data_to_fill, data_to_train, random_forest, "value_eur", save=True
);

model_name_RandomForestRegressor_value_eur_2022-06-21.pkl
Score do modelo: 0.9810


### Preenchendo os dados a partir do modelo criado

In [44]:
data.loc[players_with_club_and_no_value.index, "value_eur"] = players_with_club_and_no_value.value_eur
data.loc[players_with_club_and_no_value.index]

sofifa_id                                         player_url  \
11929     115909  https://sofifa.com/player/115909/ruben-castro-...   
11987     124344  https://sofifa.com/player/124344/ibrahim-oztur...   
12264     153066  https://sofifa.com/player/153066/lucas-licht/2...   
13149     152912  https://sofifa.com/player/152912/jose-sand/220002   
13177     110381  https://sofifa.com/player/110381/maximiliano-r...   
14103     262371  https://sofifa.com/player/262371/craig-dundas/...   
14611     140222  https://sofifa.com/player/140222/paulo-da-silv...   
15125     138703  https://sofifa.com/player/138703/andres-daless...   
16065     158293  https://sofifa.com/player/158293/zhi-zheng/220002   
16498       2702  https://sofifa.com/player/2702/kevin-ellison/2...   
16814      29118  https://sofifa.com/player/29118/shinji-ono/220002   
18159      29552  https://sofifa.com/player/29552/shunsuke-nakam...   
18166     254704  https://sofifa.com/player/254704/kazuyoshi-miu...   

            short_name                     long_name player_positions  \
11929     Rubén Castro           Rubén Castro Martín               ST   
11987        İ. Öztürk                İbrahim Öztürk               CB   
12264         L. Licht            Lucas Matías Licht      LB, LWB, LM   
13149          J. Sand             José Gustavo Sand               ST   
13177     M. Rodríguez   Maximiliano Rubén Rodríguez       LW, LM, ST   
14103        C. Dundas                  Craig Dundas               ST   
14611      P. Da Silva  Paulo César Da Silva Barrios               CB   
15125  A. D'Alessandro   Andrés Nicolás D'Alessandro          CAM, RM   
16065        Zheng Zhi                            郑智          CDM, CM   
16498       K. Ellison           Kevin Keith Ellison           LM, ST   
16814           S. Ono                         小野 伸二          CAM, CM   
18159      S. Nakamura                         中村 俊輔      CAM, CF, RW   
18166         K. Miura                         三浦 知良               ST   

       overall  potential  value_eur  wage_eur         dob  weight_kg  \
11929       69         69   957500.0    3000.0  1981-06-27         68   
11987       67         67   825750.0    3000.0  1981-06-21         80   
12264       69         69   919000.0    5000.0  1981-04-06         72   
13149       75         75  3979000.0   13000.0  1980-07-17         79   
13177       73         73  2323250.0   10000.0  1981-01-02         74   
14103       56         56   184950.0    1000.0  1981-02-16         89   
14611       67         67   838750.0     500.0  1980-02-01         76   
15125       75         75  3822000.0     900.0  1981-04-15         69   
16065       67         67   845750.0    5000.0  1980-08-20         75   
16498       58         58   262750.0    2000.0  1979-02-23         76   
16814       58         58   250550.0     800.0  1979-09-27         76   
18159       63         63   501250.0     800.0  1978-06-24         71   
18166       59         59   274250.0     700.0  1967-02-26         72   

       height_cm  club_team_id club_position  club_jersey_number  \
11929        169      100851.0            ST                 7.0   
11987        186      101006.0            CB                38.0   
12264        174      101084.0           SUB                25.0   
13149        182      110395.0            LS                 9.0   
13177        173      110396.0           SUB                11.0   
14103        188      110799.0           RES                14.0   
14611        180      111007.0           LCB                 3.0   
15125        174      111325.0           SUB                10.0   
16065        180      111839.0           LCM                10.0   
16498        185      112254.0           SUB                22.0   
16814        175      112444.0           SUB                44.0   
18159        178      113197.0           RES                10.0   
18166        177      113197.0           RES                11.0   

      preferr

In [45]:
count_na(data)

value_eur                61
wage_eur                 61
club_team_id             61
club_position            61
club_jersey_number       61
release_clause_eur     1176
player_tags           17798
player_traits          9841
pace                   2132
shooting               2132
passing                2132
dribbling              2132
defending              2132
physic                 2132
goalkeeping_speed     17107
dtype: int64

In [46]:
# Removendo jogadores não ativos
data = data.query("not club_team_id.isna()").copy()

count_na(data)

release_clause_eur     1115
player_tags           17744
player_traits          9823
pace                   2124
shooting               2124
passing                2124
dribbling              2124
defending              2124
physic                 2124
goalkeeping_speed     17054
dtype: int64

# Prevendo pace, shooting, passing, dribbling, defending, physic dos goleiros

In [47]:
gk_data = data.query("GK == 1").copy()
data_without_gk = data.query("GK == 0").copy()

columns_to_predict = ["pace", "shooting", "passing", "dribbling", "defending", "physic"]

for column_to_predict in columns_to_predict:

    columns_to_drop_with_other_atributes = [col for col in columns_to_predict if col != column_to_predict]
    columns_to_drop_with_other_atributes.extend(columns_to_drop)

    data_to_train = data_without_gk.drop(columns=columns_to_drop_with_other_atributes)
    data_to_predict = gk_data.drop(columns=columns_to_drop_with_other_atributes)

    linear_regression = LinearRegression()
    print(f"Prevendo {column_to_predict}")
    filled_data = predict_and_populate(
        data_to_predict, data_to_train, linear_regression,
        column_to_predict, 0, save=True
    )
    print()

    data.loc[filled_data.index, column_to_predict] = filled_data[column_to_predict]

# data.loc[gk_data.index]

Prevendo pace
model_name_LinearRegression_pace_2022-06-21.pkl
Score do modelo: 0.9993

Prevendo shooting
model_name_LinearRegression_shooting_2022-06-21.pkl
Score do modelo: 0.9991

Prevendo passing
model_name_LinearRegression_passing_2022-06-21.pkl
Score do modelo: 0.9961

Prevendo dribbling
model_name_LinearRegression_dribbling_2022-06-21.pkl
Score do modelo: 0.9991

Prevendo defending
model_name_LinearRegression_defending_2022-06-21.pkl
Score do modelo: 0.9997

Prevendo physic
model_name_LinearRegression_physic_2022-06-21.pkl
Score do modelo: 0.9992



In [48]:
count_na(data)

release_clause_eur     1115
player_tags           17744
player_traits          9823
goalkeeping_speed     17054
dtype: int64

## Prevendo goalkeeping_speed dos jogadores

In [49]:
players_data = data.query("GK == 0").copy()
players_data

sofifa_id                                         player_url  \
2185      188567  https://sofifa.com/player/188567/pierre-emeric...   
2186      209989  https://sofifa.com/player/209989/thomas-partey...   
2187      193301  https://sofifa.com/player/193301/alexandre-lac...   
2188      222665  https://sofifa.com/player/222665/martin-odegaa...   
2189      226110  https://sofifa.com/player/226110/nicolas-pepe/...   
...          ...                                                ...   
19234     262593  https://sofifa.com/player/262593/moritz-seiffe...   
19235     262614  https://sofifa.com/player/262614/till-muschkow...   
19236     262755  https://sofifa.com/player/262755/shalva-ogbaid...   
19237     262756  https://sofifa.com/player/262756/matteo-gumane...   
19238     263236  https://sofifa.com/player/263236/yazid-heimur/...   

           short_name                                    long_name  \
2185    P. Aubameyang  Pierre-Emerick Emiliano François Aubameyang   
2186        T. Partey                                Thomas Partey   
2187     A. Lacazette                          Alexandre Lacazette   
2188      M. Ødegaard                              Martin Ødegaard   
2189          N. Pépé                                 Nicolas Pépé   
...               ...                                          ...   
19234     M. Seiffert                              Moritz Seiffert   
19235  T. Muschkowski                             Till Muschkowski   
19236     S. Ogbaidze                              Shalva Ogbaidze   
19237      M. Gumaneh                               Matteo Gumaneh   
19238       Y. Heimur                                 Yazid Heimur   

      player_positions  overall  potential   value_eur  wage_eur         dob  \
2185        ST, LW, LM       85         85  43500000.0  145000.0  1989-06-18   
2186           CM, CDM       83         83  36500000.0  115000.0  1993-06-13   
2187                ST       82         82  30000000.0  110000.0  1991-05-28   
2188           CAM, CM       82         88  49000000.0   89000.0  1998-12-17   
2189            RM, RW       81         82  30500000.0   95000.0  1995-05-29   
...                ...      ...        ...         ...       ...         ...   
19234               LW       58         66    450000.0     600.0  2000-11-04   
19235               RW       55         65    275000.0     550.0  2001-05-06   
19236               ST       55         65    275000.0     500.0  2002-08-01   
19237               ST       55         67    300000.0     500.0  2002-09-05   
19238               CM       55         66    275000.0     500.0  2002-11-18   

       weight_kg  height_cm  club_team_id club_position  club_jersey_number  \
2185          80        187           1.0            ST                14.0   
2186          77        185           1.0           RDM                 5.0   
2187          73        175           1.0           SUB                 9.0   
2188          68        178           1.0           CAM                 8.0   
2189          73        183           1.0            RM                19.0   
...          ...        ...           ...           ...                 ...   
19234         74        180      115820.0           SUB                23.0   
19235         72        182      115820.0           SUB                18.0   
19236         75        186      115820.0           RES                28.0   
19237         76        188      115820.0           SUB                21.0   
19238         74        186      115820.0           RES                15.0   

      preferred_foot  weak_foot  skill_moves  international_reputation  \
2185           Right          4            4                         3   
2186           Right          3            3                         2   
2187           Right          4            4                         3   
2188            Left          2            5                         3   
2189            Left          2  

In [50]:
data_to_predict = players_data.drop(columns=columns_to_drop)
data_to_train = gk_data.drop(columns=columns_to_drop)

column_to_predict = "goalkeeping_speed"

linear_regression = LinearRegression()
filled_data = predict_and_populate(
    data_to_predict, data_to_train, linear_regression, column_to_predict, 0, save=True
)

data.loc[filled_data.index, column_to_predict] = filled_data[column_to_predict]

model_name_LinearRegression_goalkeeping_speed_2022-06-21.pkl
Score do modelo: 0.9992


In [51]:
count_na(data)

release_clause_eur     1115
player_tags           17744
player_traits          9823
dtype: int64

In [52]:
from datetime import date
data.drop(columns=["release_clause_eur"], inplace=True)

filled_data_save_path = os.sep.join([get_project_root_path(), "preprocessing", "data", f"filled_data_{date.today()}.tsv"])
data.to_csv(filled_data_save_path, index=False, sep="\t")
filled_data_save_path

'/home/gustavo/fifa22-players-analysis/preprocessing/data/filled_data_2022-06-21.tsv'

In [53]:
count_na(data)

player_tags      17744
player_traits     9823
dtype: int64

In [54]:
data.describe().T.round()

count       mean        std     min       25%  \
sofifa_id              19178.0   231514.0    27028.0    41.0  214568.0   
overall                19178.0       66.0        7.0    47.0      61.0   
potential              19178.0       71.0        6.0    49.0      67.0   
value_eur              19178.0  2849353.0  7611306.0  9000.0  475000.0   
wage_eur               19178.0     9018.0    19470.0   500.0    1000.0   
...                        ...        ...        ...     ...       ...   
nationality Venezuela  19178.0        0.0        0.0     0.0       0.0   
nationality Vietnam    19178.0        0.0        0.0     0.0       0.0   
nationality Wales      19178.0        0.0        0.0     0.0       0.0   
nationality Zambia     19178.0        0.0        0.0     0.0       0.0   
nationality Zimbabwe   19178.0        0.0        0.0     0.0       0.0   

                            50%        75%          max  
sofifa_id              236576.0   253563.0     264640.0  
overall                    66.0       70.0         93.0  
potential                  71.0       75.0         95.0  
value_eur              975000.0  2000000.0  194000000.0  
wage_eur                 3000.0     8000.0     350000.0  
...                         ...        ...          ...  
nationality Venezuela       0.0        0.0          1.0  
nationality Vietnam         0.0        0.0          1.0  
nationality Wales           0.0        0.0          1.0  
nationality Zambia          0.0        0.0          1.0  
nationality Zimbabwe        0.0        0.0          1.0  

[323 rows x 8 columns]